runtime/CSharp/src/Atn/ParserATNSimulator.cs

```diff
@@ -2246,18 +2244,21 @@ protected DFAState AddDFAState(DFA dfa, DFAState D)
 				return D;
 			}
 
-			lock (dfa.states)
+			var dfaStates = dfa.states;
+			lock (dfaStates)
 			{
-				DFAState existing = dfa.states.Get(D);
-				if (existing != null) return existing;
+				if (dfaStates.TryGetValue(D, out var existing))
+				{
+					return existing;
+				}
 
-				D.stateNumber = dfa.states.Count;
+				D.stateNumber = dfaStates.Count;
 				if (!D.configSet.IsReadOnly)
 				{
 					D.configSet.OptimizeConfigs(this);
 					D.configSet.IsReadOnly = true;
 				}
-				dfa.states.Put(D, D);
+				dfaStates.Add(D, D);
```

Comment:
are put/add really the same thing in C#?

Reply:
C# doesn't have `Put` method, it's utility method that is going to be removed: https://github.com/antlr/antlr4/pull/3861/commits/4108810d4e94cc40e6fda8d9d139d18f4748750a#diff-a40512909ef89c597eebbddd134e6895d6ad7eae2b2c9e5748b80c85ae2e2d47L21

C# has `[]` operator that replaces the existing value and `Add` method that throws an exception if the value is already presented in the dictionary. In this code, `Add` method is more semantically correct since checking on the existing value is written above:

```cs
if (dfaStates.TryGetValue(D, out var existing))
{
    return existing;
}
```